### PR TITLE
Upload and Download of Prometheus configuration file

### PR DESCRIPTION
**Pull request by rhsakarpos**
**What this PR does / why we need it**:
This PR enables us to 

1. Upload a configuration file to Prometheus
2. Download the configuration file from Prometheus

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
NA

**Special notes for your reviewer**:
The similar API is going to be used to upload and download configuration files for all telemetry 3rd party components, like Prometheus, Alert manager, Node exporter etc. The general flow is as below

*Upload*

- Backup the current configuration file with extension .bak
- Upload the configuration file to the server, in the right location
- Call some API or POST request to reload the configuration (if the 3rd party supports it)

**Release note**:

**Upload API**
```
http://localhost:50040/v1beta/metrics/uploadconf?conftype=prometheus
// the value of the conftype determines the component, like 'prometheus', 'nodeexporter' etc.
```

**Download API**
```
http://localhost:50040/v1beta/metrics/downloadconf?conftype=prometheus
// the value of the conftype determines the component, like 'prometheus', 'nodeexporter' etc.
```

### CODE REVIEW

**Comment by rhsakarpos** [comment]
@skdwriting @wisererik @xing-yang Please review

**Comment by rhsakarpos** [comment]
This is tested for Prometheus, am now extending it to the other 3rd party components too...that is a WIP

File: pkg/api/routers/fileops.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+func init() {
+
+	// add router for file ops api
+	filens :=
+		beego.NewNamespace("/"+constants.APIVersion,
+
+			// All operations of file can be used for both admin and users.
+			beego.NSRouter("/uploadconf", controllers.NewFileOpsPortal(), "post:UploadConfFile"),
+			beego.NSRouter("/downloadconf", controllers.NewFileOpsPortal(), "get:DownloadConfFile"),
```
Comment (xing-yang) [inline pkg/api/routers/fileops.go]: Why do we need APIs for uploading/downloading config files?

Comment (rhsakarpos) [reply pkg/api/routers/fileops.go]: This was added to add functionality to telemetry after the initial installation. Some use cases are

- add a new exporter (CEPH) to Prometheus 
- define new threshold rules in Prometheus to raise Alerts to Alert manager
etc. 
The primary driver for adding these API was that during a demo (in a meetup, conference etc.) we do NOT need to keep switching between the UI and the backend console to do these configuration changes (either by running a script or by directly editing the configuration files)

File: pkg/api/controllers/fileops.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	"os/exec"
+
+	log "github.com/golang/glog"
+
+	"github.com/opensds/opensds/pkg/controller/client"
+	"github.com/opensds/opensds/pkg/model"
+)
+
+// prometheus constants
+var PROMETHEUS_CONF_HOME string
```
Comment (leonwanghui) [inline pkg/api/controllers/fileops.go]: Please define the variable in this format like `PrometheusCONFHome`, also this style would be better:
```go
var (
    PrometheusCONFHome string
    PrometheusURL string
)
```

File: pkg/api/controllers/fileops.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	}
+}
+
+type FileOpsPortal struct {
+	BasePortal
+
+	CtrClient client.Client
+}
+
+func (this *FileOpsPortal) UploadConfFile() {
```
Comment (leonwanghui) [inline pkg/api/controllers/fileops.go]: Please don't use `this` as the pointer value, change it to `f`.

File: pkg/api/controllers/fileops.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+	m, _ := this.GetParameters()
+	confType := m["conftype"][0]
+
+	switch confType {
+	case "prometheus":
+		{
```
Comment (leonwanghui) [inline pkg/api/controllers/fileops.go]: There is no need to add this big parantheses.

File: pkg/api/controllers/fileops.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+func (this *FileOpsPortal) DownloadConfFile() {
+
+	m, _ := this.GetParameters()
+	confType := m["conftype"][0]
+
+	switch confType {
+	case "prometheus":
+		{
```
Comment (leonwanghui) [inline pkg/api/controllers/fileops.go]: Same here

File: pkg/api/routers/fileops.go (hunk 1 above)
Comment (leonwanghui) [reply pkg/api/routers/fileops.go]: I don't think it's a good idea to add a new first-level API, if this is a required operation, please move it under metrics path.

File: pkg/api/controllers/fileops.go (hunk 2 above)
Comment (rhsakarpos) [reply pkg/api/controllers/fileops.go]: refactored as per IDE suggestion, to following
var PrometheusConfHome string
var PrometheusUrl string
var PrometheusConfFile string

// alert manager constants
var AlertmgrConfHome string
var AlertmgrUrl string
var AlertmgrConfFile string

var GrafanaConfHome string
var GrafanaRestartCmd string
var GrafanaConfFile string

var ReloadPath string
var BackupExtension string

File: pkg/api/controllers/fileops.go (hunk 3 above)
Comment (rhsakarpos) [reply pkg/api/controllers/fileops.go]: changed

File: pkg/api/controllers/fileops.go (hunk 4 above)
Comment (rhsakarpos) [reply pkg/api/controllers/fileops.go]: removed

File: pkg/api/controllers/fileops.go (hunk 5 above)
Comment (rhsakarpos) [reply pkg/api/controllers/fileops.go]: removed

File: pkg/api/routers/fileops.go (hunk 1 above)
Comment (rhsakarpos) [reply pkg/api/routers/fileops.go]: added subpath ```metrics/```

File: pkg/api/controllers/fileops.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+	// save file to disk
+	fSaveErr := fileOpsPortal.SaveToFile("conf_file", path)
+	if fSaveErr != nil {
+		log.Errorf("error saving file %s", path)
+	} else {
+		if toCallReloadEndpoint == true {
+			reloadResp, reloadErr := http.Post(url+reloadPath, "application/json", nil)
+			if reloadErr != nil {
+				log.Errorf("Error on reload of configuration %s", reloadErr)
```
Comment (xing-yang) [inline pkg/api/controllers/fileops.go]: Start the msg with lower case.

File: pkg/api/controllers/fileops.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+				fileOpsPortal.ErrorHandle(model.ErrorInternalServer, reloadErr.Error())
+				return
+			}
+			respBody, readBodyErr := ioutil.ReadAll(reloadResp.Body)
+			if readBodyErr != nil {
+				log.Errorf("Error on reload of configuration %s", reloadErr)
```
Comment (xing-yang) [inline pkg/api/controllers/fileops.go]: Start the msg with lower case.

File: pkg/api/controllers/fileops.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2019 OpenSDS Authors All Rights Reserved.
```
Comment (leonwanghui) [inline pkg/api/controllers/fileops.go]: Please remove `All Rights Reserved`

**Comment by rhsakarpos** [comment]
> I mean can you move the code in `fileops.go` to `metrics.go`. Because downloadconf and uploadconf is subpath under metrics.

refactored, moved into metrics.go

File: pkg/api/controllers/fileops.go (hunk 8 above)
Comment (rhsakarpos) [reply pkg/api/controllers/fileops.go]: removed

File: pkg/api/controllers/fileops.go (hunk 6 above)
Comment (rhsakarpos) [reply pkg/api/controllers/fileops.go]: updated.

File: pkg/api/controllers/fileops.go (hunk 7 above)
Comment (rhsakarpos) [reply pkg/api/controllers/fileops.go]: updated

**Comment by leonwanghui** [comment]
Please rebase it.

**Comment by rhsakarpos** [comment]
> Please rebase it.

Done

File: pkg/api/routers/metric.go [hunk 9]
```diff
@@ -22,6 +22,16 @@ import (
 
 func init() {
 
+	// add router for file ops api
+	filens :=
+		beego.NewNamespace("/"+constants.APIVersion,
```
Comment (leonwanghui) [inline pkg/api/routers/metric.go]: Please append it in the existing namespace.

Comment (rhsakarpos) [reply pkg/api/routers/metric.go]: done

File: pkg/api/routers/metric.go [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 OpenSDS Authors.
+// Copyright (c) 2019 OpenSDS Authors
```
Comment (leonwanghui) [inline pkg/api/routers/metric.go]: Please add `.`

**Comment by leonwanghui** [comment]
Also need a rebase.

File: pkg/api/controllers/metrics.go [hunk 11]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 The OpenSDS Authors All Rights Reserved.
+// Copyright (c) 2019 The OpenSDS Authors
```
[the remark this answers is not in the document]
Comment (rhsakarpos) [reply pkg/api/controllers/metrics.go]: added

File: pkg/api/routers/metric.go (hunk 10 above)
Comment (rhsakarpos) [reply pkg/api/routers/metric.go]: added

File: pkg/api/controllers/metrics.go (hunk 11 above)
Comment (leonwanghui) [inline pkg/api/controllers/metrics.go]: Still missing here.

File: pkg/api/routers/metric.go (hunk 10 above)
Comment (leonwanghui) [inline pkg/api/routers/metric.go]: Still missing here.

**Comment by leonwanghui** [comment]
@wisererik Please take a look at this PR.

**Pull request merged**